6 editing/caret/ios tests consistently fail on some recent simulator versions

whsieh opened the pull request:
#### 6723dcfd85e9413043566d211736078ffb289878
<pre>
6 editing/caret/ios tests consistently fail on some recent simulator versions
<a href="https://bugs.webkit.org/show_bug.cgi?id=256018">https://bugs.webkit.org/show_bug.cgi?id=256018</a>
rdar://108559805

Reviewed by Aditya Keerthi.

Use `UITextSelectionDisplayInteraction` to identify ranged selection views, caret views, and
selection handles when appropriate. See changes below for more details.

* LayoutTests/editing/caret/ios/absolute-caret-position-after-scroll.html:
* LayoutTests/editing/caret/ios/caret-in-overflow-area.html:
* LayoutTests/editing/caret/ios/emoji.html:
* LayoutTests/editing/caret/ios/fixed-caret-position-after-scroll.html:
* LayoutTests/editing/selection/character-granularity-rect.html:
* LayoutTests/editing/selection/ios/absolute-selection-after-scroll.html:
* LayoutTests/editing/selection/ios/fixed-selection-after-scroll.html:
* LayoutTests/editing/selection/ios/selection-extends-into-overflow-area.html:
* LayoutTests/fast/events/touch/ios/long-press-then-drag-up-to-change-selected-text-overflow.html:
* LayoutTests/fast/events/touch/ios/selection-handles-after-touch-end.html:

Refactor these existing tests, so that they use `selectionCaretViewRect` rather than
`textSelectionCaretRect` and `selectionRangeViewRects` instead of `textSelectionRangeRects`. There&apos;s
no reason to keep both versions of these testing hooks working, so I&apos;m making these tests all use
the versions that check the platform `UIView` geometry directly, which better reflects what the
user sees.

* LayoutTests/fast/forms/textarea/ios/caret-x-position-in-textarea-matches-textfield.html:
* LayoutTests/fast/visual-viewport/ios/caret-after-focus-in-fixed.html:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getUICaretRect.return.new.Promise.): Deleted.
(window.UIHelper.getUICaretRect.return.new.Promise): Deleted.
(window.UIHelper.getUICaretRect): Deleted.
(window.UIHelper.getUISelectionRects.return.new.Promise.): Deleted.
(window.UIHelper.getUISelectionRects.return.new.Promise): Deleted.
(window.UIHelper.getUISelectionRects): Deleted.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _uiTextCaretRect]):

The changes above also allow us to remove this SPI hook (though, I&apos;m still leaving behind a method
stub for now, for binary compatibility).

* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _uiTextSelectionRects]): Deleted.
(-[WKWebView _serializedSelectionCaretBackgroundColorForTesting]): Deleted.

Remove these testing hooks altogether, since they&apos;re defined only in the testing header. The former
is no longer necessary since we now only use `selectionRangeViewRects` in the test runner, and the
latter has been moved into `UIScriptController`, out of engine code.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _uiTextSelectionRects]): Deleted.
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:

Apply a similar treatment to several API tests that run into the same issue.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/testing/cocoa/CocoaColorSerialization.h: Added.
* Source/WebCore/testing/cocoa/CocoaColorSerialization.mm: Added.
(WebCoreTestSupport::serializationForCSS):

Add a testing hook to convert platform Cocoa colors (`NSColor`, `UIColor`) to a serialized string
for CSS. We use this in `UIScriptControllerIOS` below.

(TEST):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView caretViewRectInContentCoordinates]):
(-[TestWKWebView selectionViewRectsInContentCoordinates]):
(-[TestWKWebView textSelectionDisplayInteraction]):
(-[TestWKWebView selectionRectsAfterPresentationUpdate]): Deleted.
* Tools/WebKitTestRunner/ios/PlatformWebViewIOS.mm:
(WTR::PlatformWebView::windowSnapshotImage):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::notifyDone):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::selectionStartGrabberViewRect const):
(WTR::UIScriptControllerIOS::selectionEndGrabberViewRect const):

Search for `_UITextSelectionLollipopView` in the view hierarchy, for now. `-handleViews` seems to
return an empty array, so we can&apos;t adopt it quite yet.

(WTR::UIScriptControllerIOS::selectionCaretViewRect const):
(WTR::UIScriptControllerIOS::selectionRangeViewRects const):

In the case where `selectionView` is nil but `HAVE(UI_TEXT_SELECTION_DISPLAY_INTERACTION)` is set,
fall back to asking for `-cursorView` and `-highlightView`, respectively. Note that the `-hidden`
check is necessary, since the view bounds remain at their last set value even when the selection is
deactivated; instead, the caret, selection highlights and selection handles are simply hidden when
not active.

(WTR::UIScriptControllerIOS::selectionCaretBackgroundColor const):
(WTR::UIScriptControllerIOS::textSelectionDisplayInteraction const):
(WTR::UIScriptControllerIOS::textSelectionRangeRects const): Deleted.
(WTR::UIScriptControllerIOS::textSelectionCaretRect const): Deleted.

Canonical link: <a href="https://commits.webkit.org/263460@main">https://commits.webkit.org/263460@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8714560a12cf49173dac9d80a5b212667773ec44

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4693 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4813 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4973 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6195 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4844 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4686 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4961 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4778 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5072 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4769 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4856 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4201 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6205 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2345 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4192 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9175 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4204 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4266 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5829 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4673 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3809 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4188 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1148 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
```diff
@@ -1295,11 +1320,25 @@ static UIDeviceOrientation toUIDeviceOrientation(DeviceOrientation orientation)
+    if (alpha == 1)
+        colorDescription = [NSString stringWithFormat:@"rgb(%.0f, %.0f, %.0f)", floor(red * 255), floor(green * 255), floor(blue * 255)];
+    else
+        colorDescription = [NSString stringWithFormat:@"rgba(%.0f, %.0f, %.0f, %.3f)", floor(red * 255), floor(green * 255), floor(blue * 255), alpha];
```
Can we reuse `WebCore/ColorSerialization.h` and `WebCore/ColorCocoa.h` here?

whsieh replied:
I tried that at first, but unfortunately ran into some build failures when attempting to link WebCore from WebKitTestRunner. I think it would be doable by exposing a WebCoreTestSupport wrapper instead though, so I'll try this approach and see if it works.

pxlcoder commented on Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
```diff
@@ -746,34 +746,34 @@ - (void)evaluateJavaScriptAndWaitForInputSessionToChange:(NSString *)script
-    NSArray<UITextSelectionRect *> *rects = [self.textInputContentView valueForKeyPath:@"interactionAssistant.selectionView.rangeView.rects"];
-    for (UITextSelectionRect *rect in rects)
-        [selectionRects addObject:[NSValue valueWithCGRect:rect.rect]];
+    NSArray *rects = [self.textInputContentView valueForKeyPath:@"interactionAssistant.selectionView.rangeView.rects"];
```
Can't we keep the type as `NSArray<UITextSelectionRect *>`, and use `UITextSelectionRect` (and the `rect` property) in the loop below?

whsieh replied:
Yes — good catch.

pxlcoder reviewed (approved):
Would you mind seeing if `fast/events/touch/ios/double-tap-on-editable-content-for-selection-then-drag-up-to-change-selected-text.html` passes following this change?

pxlcoder commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
```diff
@@ -860,8 +856,19 @@ static void clipSelectionViewRectToContentView(CGRect& rect, UIView *contentView
+
+#if HAVE(UI_TEXT_SELECTION_DISPLAY_INTERACTION)
+    if (!handleView) {
+        // FIXME: We should be able to use -handleViews here, but this seems returns an empty array,
```
This one is probably worth a UIKit bug.

whsieh replied:
Good point! Filed rdar://108607881, and I'll link it here.

whsieh committed:
6 editing/caret/ios tests consistently fail on some recent simulator versions
https://bugs.webkit.org/show_bug.cgi?id=256018
rdar://108559805

Reviewed by Aditya Keerthi.

Use `UITextSelectionDisplayInteraction` to identify ranged selection views, caret views, and
selection handles when appropriate. See changes below for more details.

* LayoutTests/editing/caret/ios/absolute-caret-position-after-scroll.html:
* LayoutTests/editing/caret/ios/caret-in-overflow-area.html:
* LayoutTests/editing/caret/ios/emoji.html:
* LayoutTests/editing/caret/ios/fixed-caret-position-after-scroll.html:
* LayoutTests/editing/selection/character-granularity-rect.html:
* LayoutTests/editing/selection/ios/absolute-selection-after-scroll.html:
* LayoutTests/editing/selection/ios/fixed-selection-after-scroll.html:
* LayoutTests/editing/selection/ios/selection-extends-into-overflow-area.html:
* LayoutTests/fast/events/touch/ios/long-press-then-drag-up-to-change-selected-text-overflow.html:
* LayoutTests/fast/events/touch/ios/selection-handles-after-touch-end.html:

Refactor these existing tests, so that they use `selectionCaretViewRect` rather than
`textSelectionCaretRect` and `selectionRangeViewRects` instead of `textSelectionRangeRects`. There's
no reason to keep both versions of these testing hooks working, so I'm making these tests all use
the versions that check the platform `UIView` geometry directly, which better reflects what the
user sees.

* LayoutTests/fast/forms/textarea/ios/caret-x-position-in-textarea-matches-textfield.html:
* LayoutTests/fast/visual-viewport/ios/caret-after-focus-in-fixed.html:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getUICaretRect.return.new.Promise.): Deleted.
(window.UIHelper.getUICaretRect.return.new.Promise): Deleted.
(window.UIHelper.getUICaretRect): Deleted.
(window.UIHelper.getUISelectionRects.return.new.Promise.): Deleted.
(window.UIHelper.getUISelectionRects.return.new.Promise): Deleted.
(window.UIHelper.getUISelectionRects): Deleted.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _uiTextCaretRect]):

The changes above also allow us to remove this SPI hook (though, I'm still leaving behind a method
stub for now, for binary compatibility).

* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(-[WKWebView _uiTextSelectionRects]): Deleted.
(-[WKWebView _serializedSelectionCaretBackgroundColorForTesting]): Deleted.

Remove these testing hooks altogether, since they're defined only in the testing header. The former
is no longer necessary since we now only use `selectionRangeViewRects` in the test runner, and the
latter has been moved into `UIScriptController`, out of engine code.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _uiTextSelectionRects]): Deleted.
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:

Apply a similar treatment to several API tests that run into the same issue.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/testing/cocoa/CocoaColorSerialization.h: Added.
* Source/WebCore/testing/cocoa/CocoaColorSerialization.mm: Added.
(WebCoreTestSupport::serializationForCSS):

Add a testing hook to convert platform Cocoa colors (`NSColor`, `UIColor`) to a serialized string
for CSS. We use this in `UIScriptControllerIOS` below.

(TEST):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView caretViewRectInContentCoordinates]):
(-[TestWKWebView selectionViewRectsInContentCoordinates]):
(-[TestWKWebView textSelectionDisplayInteraction]):
(-[TestWKWebView selectionRectsAfterPresentationUpdate]): Deleted.
* Tools/WebKitTestRunner/ios/PlatformWebViewIOS.mm:
(WTR::PlatformWebView::windowSnapshotImage):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::notifyDone):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::selectionStartGrabberViewRect const):
(WTR::UIScriptControllerIOS::selectionEndGrabberViewRect const):

Search for `_UITextSelectionLollipopView` in the view hierarchy, for now. `-handleViews` seems to
return an empty array, so we can't adopt it quite yet.

(WTR::UIScriptControllerIOS::selectionCaretViewRect const):
(WTR::UIScriptControllerIOS::selectionRangeViewRects const):

In the case where `selectionView` is nil but `HAVE(UI_TEXT_SELECTION_DISPLAY_INTERACTION)` is set,
fall back to asking for `-cursorView` and `-highlightView`, respectively. Note that the `-hidden`
check is necessary, since the view bounds remain at their last set value even when the selection is
deactivated; instead, the caret, selection highlights and selection handles are simply hidden when
not active.

(WTR::UIScriptControllerIOS::selectionCaretBackgroundColor const):
(WTR::UIScriptControllerIOS::textSelectionDisplayInteraction const):
(WTR::UIScriptControllerIOS::textSelectionRangeRects const): Deleted.
(WTR::UIScriptControllerIOS::textSelectionCaretRect const): Deleted.

Canonical link: https://commits.webkit.org/263460@main
